normalize deps Makefile targets and add uninstall

vtjnash opened the pull request:
using a consistent mapping between target names reduces the coupling between targets
and should make it easier to add features like reinstall, uninstall, etc.

and it fixes #17671 and #17282 

while the diff is large, this essentially just a mechanical change.

all targets (except get) now use secondary file targets, giving a simple mapping from the PHONY targets, and allows another target to reference the manifest file directly rather than using the OBJ_TARGET variable:
extract: srccache/LIB/source-extracted
configure: bulid/LIB/build-configured
compile: build/LIB/build-compiled
check: build/LIB/build-checked
install: usr/manifest/LIB

this allows more accurate tracking of when a build stage has completed, and a predictable way for the user to force a build stage to restart (by removing or touching the corresponding file)

(also killed of the atlas target, since it hasn't been maintained in several years)

vtjnash committed:
normalize deps Makefile targets

using a consistent mapping between target names reduces the coupling between targets
and should make it easier to add features like reinstall, uninstall, etc.

and it fixes #17671 and #17282

tkelman commented on deps/tools/common.mk:
```diff
@@ -0,0 +1,93 @@
+## phony targets ##
+
+.PHONY: default compile install cleanall distcleanall \
+	get-* configure-* compile-* check-* install-* \
```
`extract` and `extract-` should probably be here?

tkelman commented on deps/Makefile:
```diff
@@ -30,89 +32,34 @@ include $(SRCDIR)/llvm-ver.make
 
-STAGE1_DEPS :=
-STAGE2_DEPS :=
+DEP_LIBS :=
```
update `contrib/windows/msys_build.sh` for this renaming or appveyor's going to have a bad time

tkelman commented on deps/blas.mk:
```diff
@@ -185,61 +131,66 @@ endif
 clean-lapack:
+	-rm -f $(build_prefix)/manifest/lapack $(BUILDDIR)/lapack-$(LAPACK_VER)/build-compiled0 $(BUILDDIR)/lapack-$(LAPACK_VER)/build-compiled
 	-$(MAKE) -C $(BUILDDIR)/lapack-$(LAPACK_VER) clean
-	-rm -f $(LAPACK_OBJ_SOURCE) $(LAPACK_OBJ_TARGET)
```
should still be deleting these

tkelman commented on deps/arpack.mk:
```diff
@@ -44,66 +37,74 @@ $(SRCDIR)/srccache/arpack-ng-$(ARPACK_VER).tar.gz: | $(SRCDIR)/srccache
 	done
 endif
-	touch -c $@
+	echo 1 > $@
```
version?

tkelman commented on deps/libgit2.mk:
```diff
@@ -76,14 +75,15 @@ ifeq ($(OS),Linux)
 	done
 endif
-	touch -c $@
+	echo $(LIBGIT2_VER) > $@
```
sha?

tkelman commented on deps/objconv.mk:
```diff
@@ -1,28 +1,32 @@
+
+$(build_prefix)/manifest/objconv: $(BUILDDIR)/objconv/build-compiled | $(build_depsbindir) $(build_prefix)/manifest
+	cp -f $(BUILDDIR)/objconv/objconv $(build_depsbindir)/objconv
+	echo > $@
```
something? 1?

vtjnash replied:
it looks like the current version is 2.42

tkelman replied:
he doesn't version his downloads from what I remember

tkelman commented on deps/fftw.mk:
```diff
@@ -86,20 +73,18 @@ else ifeq ($(OS), Linux)
-	-$(MAKE) -C $(BUILDDIR)/fftw-$(FFTW_VER)-double clean
-	-rm -f $(FFTW_DOUBLE_OBJ_TARGET)
+clean-fftw-%:
+	-rm -f $(build_prefix)/manifest/fftw-% $(BUILDDIR)/fftw-$(FFTW_VER)-%/build-compiled $(BUILDDIR)/fftw-$(FFTW_VER)-%/build-configured
```
and the shared library

tkelman commented on deps/fftw.mk:
```diff
@@ -63,17 +58,9 @@ else ifeq ($(OS), Linux)
 		[ -L $$filename ] || $(PATCHELF_BIN) --set-rpath '$$ORIGIN' $$filename ;\
```
collapsed right above here, you're moving the dll to `$@`, not right

tkelman commented on deps/fftw.mk:
```diff
@@ -86,20 +73,18 @@ else ifeq ($(OS), Linux)
 		[ -L $$filename ] || $(PATCHELF_BIN) --set-rpath '$$ORIGIN' $$filename ;\
```
and here, moving the dll to `$@` is bad

tkelman commented on deps/gmp.mk:
```diff
@@ -13,45 +10,53 @@ endif
 clean-gmp:
+	-rm -f $(build_prefix)/manifest/gmp $(BUILDDIR)/gmp-$(GMP_VER)/build-configured $(BUILDDIR)/gmp-$(GMP_VER)/build-compiled
 	-$(MAKE) -C $(BUILDDIR)/gmp-$(GMP_VER) clean
-	-rm -f $(GMP_OBJ_TARGET)
```
should still delete this

tkelman commented on deps/arpack.mk:
```diff
@@ -44,66 +37,74 @@ $(SRCDIR)/srccache/arpack-ng-$(ARPACK_VER).tar.gz: | $(SRCDIR)/srccache
 
 ifeq ($(USE_ATLAS), 1)
-$(BUILDDIR)/arpack-ng-$(ARPACK_VER)/config.status: | $(ATLAS_OBJ_TARGET)
+$(BUILDDIR)/arpack-ng-$(ARPACK_VER)/build-configured: | $(build_prefix)/manifest/atlas
```
delete?

tkelman commented on deps/libgit2.mk:
```diff
@@ -76,14 +75,15 @@ ifeq ($(OS),Linux)
 clean-libgit2:
+	-rm -f $(build_prefix)/manifest/libgit2 $(BUILDDIR)/$(LIBGIT2_SRC_DIR)/build-configured $(BUILDDIR)/$(LIBGIT2_SRC_DIR)/build-compiled
 	-$(MAKE) -C $(BUILDDIR)/$(LIBGIT2_SRC_DIR) clean
-	-rm -f $(LIBGIT2_OBJ_TARGET)
```
should still delete this

tkelman commented on deps/libssh2.mk:
```diff
@@ -2,10 +2,9 @@
-$(eval $(call git-external,libssh2,LIBSSH2,CMakeLists.txt,build/libssh2.$(SHLIB_EXT),$(SRCDIR)/srccache))
+$(eval $(call git-external,libssh2,LIBSSH2,,,$(SRCDIR)/srccache))
 
 LIBSSH2_OBJ_SOURCE := $(BUILDDIR)/$(LIBSSH2_SRC_DIR)/src/libssh2.$(SHLIB_EXT)
```
deleting the rest of these, but not here?

tkelman commented on deps/libssh2.mk:
```diff
@@ -23,44 +22,46 @@ ifeq ($(OS),Linux)
 	echo 1 > $@
 
-$(BUILDDIR)/$(LIBSSH2_SRC_DIR)/Makefile: $(SRCDIR)/srccache/$(LIBSSH2_SRC_DIR)/CMakeLists.txt $(SRCDIR)/srccache/$(LIBSSH2_SRC_DIR)/libssh2-mbedtls.patch-applied $(SRCDIR)/srccache/$(LIBSSH2_SRC_DIR)/libssh2-encryptedpem.patch-applied $(MBEDTLS_OBJ_TARGET)
+$(BUILDDIR)/$(LIBSSH2_SRC_DIR)/build-configured: $(SRCDIR)/srccache/$(LIBSSH2_SRC_DIR)/source-extracted $(SRCDIR)/srccache/$(LIBSSH2_SRC_DIR)/libssh2-mbedtls.patch-applied $(SRCDIR)/srccache/$(LIBSSH2_SRC_DIR)/libssh2-encryptedpem.patch-applied $(MBEDTLS_OBJ_TARGET)
```
you've deleted `MBEDTLS_OBJ_TARGET`

tkelman commented on deps/libssh2.mk:
```diff
@@ -23,44 +22,46 @@ ifeq ($(OS),Linux)
 clean-libssh2:
+	-rm -f $(build_prefix)/manifest/libssh2 $(BUILDDIR)/$(LIBSSH2_SRC_DIR)/build-configured $(BUILDDIR)/$(LIBSSH2_SRC_DIR)/build-compiled
 	-$(MAKE) -C $(BUILDDIR)/$(LIBSSH2_SRC_DIR) clean
-	-rm -f $(LIBSSH2_OBJ_TARGET)
```
should still delete this

tkelman commented on deps/llvm.mk:
```diff
@@ -522,33 +517,31 @@ else
+	echo $(LLVM_VER) > $@
 
 reinstall-llvm:
-	-rm $(LLVM_OBJ_TARGET)
```
should still delete this

tkelman commented on deps/llvm.mk:
```diff
@@ -522,33 +517,31 @@ else
 clean-llvm:
+	-rm -f $(build_prefix)/manifest/llvm $(LLVM_BUILDDIR_withtype)/build-configured $(LLVM_BUILDDIR_withtype)/build-compiled
 	-$(MAKE) -C $(LLVM_BUILDDIR_withtype) clean
-	-rm -f $(build_depsbindir)/llvm-config
```
still delete this, or run `make uninstall`

tkelman commented on deps/mbedtls.mk:
```diff
@@ -66,20 +63,23 @@ endif
+		$(BUILDDIR)/mbedtls-$(MBEDTLS_VER)/build-configured \
+		$(BUILDDIR)/mbedtls-$(MBEDTLS_VER)/build-compiled
 	-$(MAKE) -C $(BUILDDIR)/mbedtls-$(MBEDTLS_VER) clean
-	-rm -f $(MBEDTLS_OBJ_TARGET)
```
should still delete this

tkelman commented on deps/mpfr.mk:
```diff
@@ -29,38 +27,44 @@ endif
 clean-mpfr:
+	-rm -f $(build_prefix)/manifest/mpfr $(BUILDDIR)/mpfr-$(MPFR_VER)/build-configured $(BUILDDIR)/mpfr-$(MPFR_VER)/build-compiled
 	-$(MAKE) -C $(BUILDDIR)/mpfr-$(MPFR_VER) clean
-	-rm -f $(MPFR_OBJ_TARGET)
```
should still delete this

tkelman commented on deps/openlibm.mk:
```diff
@@ -1,27 +1,26 @@
 clean-openlibm:
+	-rm $(build_prefix)/manifest/openlibm $(BUILDDIR)/$(OPENLIBM_SRC_DIR)/build-compiled $(build_libdir)/libopenlibm.a
 	-$(MAKE) -C $(BUILDDIR)/$(OPENLIBM_SRC_DIR) distclean $(OPENLIBM_FLAGS)
-	-rm $(OPENLIBM_OBJ_TARGET)
```
should still delete this

tkelman commented on deps/openspecfun.mk:
```diff
@@ -1,40 +1,39 @@
 clean-openspecfun:
+	-rm $(build_prefix)/manifest/openspecfun $(BUILDDIR)/$(OPENSPECFUN_SRC_DIR)/build-compiled $(build_libdir)/libopenspecfun.a
 	-$(MAKE) -C $(BUILDDIR)/$(OPENSPECFUN_SRC_DIR) distclean $(OPENSPECFUN_FLAGS)
-	-rm $(OPENSPECFUN_OBJ_TARGET)
```
should still delete this

tkelman commented on deps/patchelf.mk:
```diff
@@ -1,42 +1,48 @@
+		$(BUILDDIR)/patchelf-$(PATCHELF_VER)/build-configured \
+		$(BUILDDIR)/patchelf-$(PATCHELF_VER)/build-compiled
 	-$(MAKE) -C $(BUILDDIR)/patchelf-$(PATCHELF_VER) clean
-	-rm -f $(PATCHELF_OBJ_TARGET)
```
this is spelled wrong, but should be deleting the installed executable

tkelman commented on deps/suitesparse.mk:
```diff
@@ -77,20 +73,23 @@ $(SUITESPARSE_OBJ_TARGET): $(SUITESPARSE_OBJ_SOURCE)
-	-$(MAKE) -C $(BUILDDIR)/SuiteSparse-$(SUITESPARSE_VER) clean
+	-rm -f $(build_prefix)/manifest/suitesparse $(BUILDDIR)/SuiteSparse-$(SUITESPARSE_VER)/build-compiled
 	-rm -fr $(BUILDDIR)/SuiteSparse-$(SUITESPARSE_VER)/lib
-	-rm -f $(SUITESPARSE_OBJ_TARGET)
```
should still delete the shared library

tkelman commented on deps/tools/git-external.mk:
```diff
@@ -33,20 +33,20 @@ $5/$1/.git/HEAD: | $$($2_SRC_FILE)/HEAD
 	@[ '$$($2_SHA1)' = "$$$$(cd $5/$1 && git show -s --format='%H' HEAD)" ] || echo $$(WARNCOLOR)'==> warning: SHA1 hash did not match $1.version file'$$(ENDCOLOR)
-	touch -c $$@
-$5/$1/$4: $5/$1/.git/HEAD
+	echo 1 > $$@
```
sha?

tkelman commented on deps/tools/git-external.mk:
```diff
@@ -56,11 +56,11 @@ $2_SRC_DIR := $1-$$($2_SHA1)
 	mkdir -p $$(dir $$@) && \
 	$(TAR) -C $$(dir $$@) --strip-components 1 -xf $$<
-	touch -c $$@
+	echo 1 > $$@
```
sha?

tkelman commented on deps/unwind.mk:
```diff
@@ -1,87 +1,93 @@
 clean-unwind:
+	-rm -f $(build_prefix)/manifest/unwind $(BUILDDIR)/libunwind-$(UNWIND_VER)/build-configured $(BUILDDIR)/libunwind-$(UNWIND_VER)/build-compiled
 	-$(MAKE) -C $(BUILDDIR)/libunwind-$(UNWIND_VER) clean
-	-rm -f $(LIBUNWIND_TARGET_OBJ) $(LIBUNWIND_TARGET_SOURCE)
```
should still delete the installed library

tkelman commented on deps/unwind.mk:
```diff
@@ -1,87 +1,93 @@
 clean-osxunwind:
+	-rm $(build_prefix)/manifest/libuv $(BUILDDIR)/libosxunwind-$(OSXUNWIND_VER)/build-compiled
 	-$(MAKE) -C $(BUILDDIR)/libosxunwind-$(OSXUNWIND_VER) clean $(OSXUNWIND_FLAGS)
-	-rm $(OSXUNWIND_OBJ_TARGET)
```
should still delete the installed shared library

tkelman commented:
You know what would do all of this boilerplate for us? `ExternalProject_Add`.

vtjnash commented:
> should still delete the shared library

it's not really that useful to delete one random file from the install, that's more a vestige of the old Makefile. in part 2 of this PR series, I'll update my PR that does a full uninstall (like autotools `make uninstall`, but functional even for handwritten makefiles and cmake)

tkelman commented:
very misleading to have `clean-foo` not actually clean anything from the installed copy of foo. otherwise you're leaving in place the piece that julia needs, so after `clean-foo` you'll have an old copy of the shared library and you won't notice that fact in julia tests

tkelman commented:
how about get a handwritten start on `uninstall` with at least the existing removals of the shared libraries

vtjnash committed:
add uninstall targets to deps

also rename deps/build to deps/scratch to make bisect easier across this
change and to reflect its new intended status as a temporary directory

vtjnash commented:
> how about get a handwritten start on uninstall with at least the existing removals of the shared libraries

in that case, it's easier just to do the full uninstall, since I've already written that code and just need to merge it here. I've fixed about half of them now.

tkelman commented on deps/pcre.mk:
```diff
@@ -11,38 +8,47 @@ endif
+
+$(eval $(call staged-install, \
+	pcre,pcre2-$$(PCRE_VER), \
+	MAKE_INSTALL,$$$$(LIBTOOL_CCLD),, \
```
I think this is causing issues on travis

tkelman commented on deps/tools/common.mk:
```diff
@@ -0,0 +1,130 @@
+
+$$(build_prefix)/manifest/$(strip $1): $$(build_staging)/$2.tgz | $(build_prefix)/manifest
+	mkdir -p $$(build_prefix)
+	tar -xUzf $$< -C $$(build_prefix)
```
and this is having issues on appveyor

vtjnash committed:
try harder to print info when deps target fails on travis

also intersperse stderr info into the log
and reduce the update interval of bar

tkelman commented on deps/tools/common.mk:
```diff
@@ -68,26 +67,71 @@ $(build_prefix): | $(DIRS)
+#gnu make
+UNTAR = $(TAR) xzf
+else
+#bsd make
```
tar

vtjnash committed:
touch old targets in srccache to make bisect easier across this change

tkelman commented:
Please do not merge this until after we've fixed up the curl build issues. This won't be a backport candidate prior to 0.5.0.

tkelman commented on deps/arpack.mk:
```diff
@@ -43,66 +48,67 @@ $(SRCDIR)/srccache/arpack-ng-$(ARPACK_VER).tar.gz: | $(SRCDIR)/srccache
+	ARPACK_INSTALL,$(ARPACK_MFLAGS),, \
+	$$(INSTALL_NAME_CMD)libarpack.$$(SHLIB_EXT) $$(build_shlibdir)/libarpack.$$(SHLIB_EXT)))
 
 clean-arpack:
```
clean and/or distclean should probably call uninstall

vtjnash replied:
that's handled automatically by staged-install

tkelman replied:
ah, great, missed that

tkelman commented on deps/libgit2.mk:
```diff
@@ -29,73 +36,74 @@ ifeq ($(OS),Linux)
-endif
-$(BUILDDIR)/$(LIBGIT2_SRC_DIR)/Makefile: $(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR)/CMakeLists.txt $(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR)/libgit2-ssh.patch-applied $(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR)/libgit2-agent-nonfatal.patch-applied
+$(BUILDDIR)/$(LIBGIT2_SRC_DIR)/build-configured: \
+	$(SRCDIR)/srccache/$(LIBGIT2_SRC_DIR)/libgit2-require-openssl.patch-applied \
```
not really needed on mac

vtjnash replied:
that's good; but the source files are target-independent

tkelman replied:
the patches might not all be written with that taken into account

vtjnash replied:
If the patches aren't in state to merge upstream, they shouldn't be in our repo either

tkelman replied:
while I agree with that, I think quite a few of the existing patches would fail that test

vtjnash replied:
Can you give an example? This is the only patch that I've come across that was attempting to be conditionally applied.

tkelman replied:
the llvm threads patch is one. the gmp one was done in a different way by upstream there. will double check the others.

tkelman commented on deps/libgit2.mk:
```diff
@@ -29,73 +36,74 @@ ifeq ($(OS),Linux)
-		[ ! -z "$$LIB_PATH" ] && cp -v -f "$$LIB_PATH" $(build_shlibdir); \
+		LIB_PATH=$$$$(echo "$$$$LIBGIT_LIBS" | grep "$$$$LIB"); \
+		echo "LIB_PATH for $$$$LIB: $$$$LIB_PATH"; \
+		[ ! -z "$$$$LIB_PATH" ] && cp -v "$$$$LIB_PATH" $$(build_shlibdir); \
```
did you remove the `-f` intentionally or as a rebase error?

vtjnash replied:
no, that's correct to remove now. but looks like I forgot to add `$2` here

tkelman commented on deps/mbedtls.mk:
```diff
@@ -25,59 +22,73 @@ endif
+ifeq ($(OS),WINNT)
+define MBEDTLS_INSTALL
+	cp $1/library/libmbedcrypto.$$(SHLIB_EXT) $2/$$(build_shlibdir)
+	cp $1/library/libmbedcrypto.$$(SHLIB_EXT) $2/$$(build_shlibdir)
```
duplicate?

tkelman commented on deps/libgit2.mk:
```diff
@@ -29,73 +36,74 @@ ifeq ($(OS),Linux)
 clean-libgit2:
+	-rm -f $(build_prefix)/manifest/libgit2 $(BUILDDIR)/$(LIBGIT2_SRC_DIR)/build-configured $(BUILDDIR)/$(LIBGIT2_SRC_DIR)/build-compiled
 	-$(MAKE) -C $(BUILDDIR)/$(LIBGIT2_SRC_DIR) clean
-	-rm -f $(build_shlibdir)/libgit2* $(build_shlibdir)/libssl* $(build_shlibdir)/libcrypto*
```
these don't go through destdir staging so won't uninstall miss them?

tkelman commented on deps/llvm.mk:
```diff
@@ -403,20 +399,20 @@ endif # LLVM_GIT_VER_POLLY
 	cd $$(LLVM_SRC_DIR) && patch -p1 < $$(SRCDIR)/patches/$1.patch
 	echo 1 > $$@
 LLVM_PATCH_PREV := $$(LLVM_SRC_DIR)/$1.patch-applied
-LLVM_PATCH_LIST += $$(LLVM_PATCH_PREV)
```
I think removing this would mean that if anyone adds a patch later on and it's not at the very end of the list, it would get skipped

vtjnash replied:
all are still dependencies (http://stackoverflow.com/a/24856711)

tkelman replied:
I think you're wrong here, as I'm only getting a single patch being applied.

vtjnash replied:
I see them all getting applied, so I can't intuit what's wrong with your setup.

tkelman replied:
I think it's an issue with incremental builds where the extracted source already exists but the tag files created by this PR don't.

tkelman commented on deps/curl.mk:
```diff
@@ -1,62 +1,63 @@
+	$$(INSTALL_NAME_CMD)libcurl.$$(SHLIB_EXT) $$(build_shlibdir)/libcurl.$$(SHLIB_EXT)))
 
 clean-curl:
+	-rm -rf $(build_prefix)/manifest/curl \
```
why do only some of the clean targets delete the manifest?

vtjnash replied:
none of them should need to, so that's just missed cleanup

vtjnash committed:
remove temporary install files to reduce disk usage slightly

vtjnash committed:
libarpack RPATH $ORIGIN is handle by configure now

patchelf bin is not required and was a no-op

vtjnash commented:
Anyone opposed to merging this right now, while the queues are quiet?

tkelman commented:
This is going to make everyone rebuild everything, the bugs it fixes are relatively minor.

vtjnash commented:
fwiw, I expect the x86_64 CI linalg/arnoldi here will fail, since I corrupted the arpack cache on my previous push. I can reset the cache, but then the PR build will take longer, and the cache shouldn't transfer to master after merging, does it?

@tkelman it fixes my out-of-tree build from rebuilding several things quite frequently, letting me use out-of-tree builds more heavily without the current conflicts.

tkelman commented:
Has this PR ever successfully finished a 32 bit build within the time limit?

tkelman commented on .travis.yml:
```diff
@@ -112,9 +128,6 @@ script:
     - cd `dirname $TRAVIS_BUILD_DIR` && mv julia2 julia &&
-        rm -rf julia/deps/build/julia-env &&
-        rm -rf julia/deps/build/libssh2-*/CMakeFiles/Makefile2 &&
-        rm -rf julia/deps/build/curl-*/config.log &&
```
are you positive these won't keep recaching?

vtjnash replied:
they shouldn't, since we should no longer be touching & re-running the rules that were causing this (install should be getting handled by usr-staging.tgz now and completely ignoring deps/scratch)

and confirmed by CI: https://travis-ci.org/JuliaLang/julia/jobs/157787509#L1631

tkelman commented on deps/utf8proc.mk:
```diff
@@ -1,34 +1,40 @@
+extract-utf8proc: $(BUILDDIR)/$(UTF8PROC_SRC_DIR)/source-extracted
+configure-utf8proc: extract-utf8proc
+compile-utf8proc: $(BUILDDIR)/$(UTF8PROC_SRC_DIR)/build-compiled
+# utf8proc tests disabled since they require a download
```
so do a bunch of other things, it's not like they take long to run

vtjnash replied:
yes, but the intent of fastcheck is to be something a buildbot could run, and they don't necessarily have network access

tkelman replied:
the only buildbots that don't have network access are the ones that use system versions of deps for nearly everything

vtjnash commented:
The 32-bit build wouldn't finish on master either if it was held to the same time limit (https://travis-ci.org/JuliaLang/julia/jobs/153542507)

tkelman commented:
That build isn't very meaningful as it was an issue on Travis' side. The from-scratch build has gone through in under 2 hours before to repopulate a clean cache, and if it doesn't on this branch then it may be making the build slower by adding more steps. Need to find a way to fix that otherwise every build will fail.

vtjnash commented:
ok, I don't know how to search for the last full cache build

vtjnash commented:
anyways, full deps build for x86 have been taking ~65min on this PR. with typical travis variance, that gives about a 20% chance of failure given that the cache build times range from 45-80 minutes. and only need one build to pass to populate the cache for a long time after.

this also implements support for doing our own cache. E.g. Our buildbots can upload their usr-staging folders and travis can use those instead, eliminating that issue and the difficulties we've had with resetting corrupted caches. Plus we can enable that for everyone, so that everyone downloading the julia src repo can get lightning fast builds.

tkelman commented:
You're neglecting the fact that compilers and glibc versions need to match to make that actually work.

tkelman commented:
Did you test this in clean compiles on all platforms?
```
/usr/bin/tar: ./bin: Cannot open: File exists
/usr/bin/tar: Exiting with failure status due to previous errors
make[1]: *** [/home/Tony/julia32/deps/dsfmt.mk:44: /home/Tony/julia32/usr/manifest/dsfmt] Error 2
```

tkelman commented:
`make distcleanall` reconfigures and recompiles libuv. This really wasn't ready to be merged, which is why when you asked "Anyone opposed to merging this right now" I responded yes.

tkelman commented:
It's also broken on the mac buildbot with `ld: library not found -lcholmod`. This needed way more testing.

tkelman commented on deps/blas.mk:
```diff
@@ -76,171 +74,105 @@ endif
+	echo 1 > $@
+
+define OPENBLAS_INSTALL
+    $(call SHLIBFILE_INSTALL,$1,$2,$3)
```
should this be a tab, just for visual consistency?

vchuravy commented:
I am compiling with llmv 3.9 and it seems that the files are not getting install in the right place:
```
tar tvf /home/wallnuss/src/julia/usr-staging/llvm-3.9.0/build_Release+Asserts.tgz
drwxr-xr-x wallnuss/wallnuss 0 2016-09-07 01:22 ./
```

I am still looking into what causes the issue, but it seems the build is ignoring the `($build_prefix)` in  https://github.com/JuliaLang/julia/blob/master/deps/tools/common.mk#L146

tkelman commented on deps/unwind.mk:
```diff
@@ -1,87 +1,101 @@
+	echo $(OSXUNWIND_VER) > $(build_prefix)/manifest/osxunwind
 
 clean-osxunwind:
+	-rm $(build_prefix)/manifest/libuv $(BUILDDIR)/libosxunwind-$(OSXUNWIND_VER)/build-compiled
```
why is this deleting the libuv manifest?

tkelman commented on deps/mbedtls.mk:
```diff
@@ -25,59 +22,72 @@ endif
+
 
 clean-mbedtls:
+	-rm -f $(build_prefix)/manifest/mbedtls \
```
should clean delete the manifest or not? (likewise for suitesparse)

vtjnash replied:
suitesparse doesn't use staged install, so it needs to

here it is unnecessary, so it shouldn't, but it also doesn't matter

tkelman replied:
these are the only places that do - so if it's not necessary in mbedtls, best to remove it for consistency / clarity

vtjnash commented:
thanks @vchuravy. fixed by https://github.com/JuliaLang/julia/pull/18378/commits/b3541bc137b69dad7c980dcb6868fdc1d7b4d338#diff-c0f89475cb4381b2b7a2e7228fe5e39cR518

vchuravy commented:
@vtjnash thanks! I got to the same fix and was just waiting for a build finishing before submitting it. This solves my problem :)
